### PR TITLE
Move everything related to cancellation to CancellableContext.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Currently the logic of cancellation is split between Context and CancellableContext where some of the logic related to "isCancelled" lives in the CancellableContext, and some of the logic related to listeners added lives in "Context".

This PR ensures that the entire logic related to cancellation lives in one place (CancellableContext) and ensures clear separation of concerns, here are some benefits:
* Clear ownership of the listeners and cancellation property.
* Reduce the size of the Context object (2 empty pointers).

Because the list of listeners is shared between all child Contexts of a "CancellableContext" it may increases contention and slows the linear scanning during removal, but it seems like it won't be a problem because listeners seem to be rare, so 0→1 and 1→0 transitions should be the most common which stresses the same code paths.

Changes in cancellation ordering:
* The overall ordering stays the same: first listeners added to the CancellableContext then child contexts.
* Important Change: in the previous implementation it was guaranteed that listeners added to a child Context will be executed in the same group, now this is no longer the case, they are executed in the same order they were added.
* In the previous implementation all child Context of a CancellableContext (including child of a child of a CancellableContext) are adding their "parentListener" to the CancellableContext list, which means that there is no guaranteed ordering of execution in the case of a parent and child Context when none of them are CancellableContext.

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Unrelated failure:
```
Downloading from central: https://repo.maven.apache.org/maven2/kr/motd/maven/os-maven-plugin/1.6.2/os-maven-plugin-1.6.2.pom
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Unresolveable build extension: Plugin kr.motd.maven:os-maven-plugin:1.6.2 or one of its dependencies could not be resolved: Failed to read artifact descriptor for kr.motd.maven:os-maven-plugin:jar:1.6.2 @ 
```

**Comment by bogdandrutu** [comment]
> Also, if that is your core objective I would point out that PersistentHashArrayMappedTrie is a useless wrapper class that could trivially be avoided by using PersistentHashArrayMappedTrie.Node directly. That would save 48 bytes with only trivial code changes.

See https://github.com/grpc/grpc-java/pull/6945

File: context/src/main/java/io/grpc/Context.java [hunk 1]
```diff
@@ -459,9 +453,8 @@ boolean isCurrent() {
   public boolean isCancelled() {
     if (cancellableAncestor == null) {
       return false;
-    } else {
-      return cancellableAncestor.isCancelled();
     }
+    return cancellableAncestor.isCancelled();
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply context/src/main/java/io/grpc/Context.java]: Agreed, reverted, I did it mostly for consistency with other methods (but not important for this PR).

File: context/src/main/java/io/grpc/Context.java [hunk 2]
```diff
@@ -498,98 +490,30 @@ public void addListener(final CancellationListener cancellationListener,
...
-    }
-    for (int i = 0; i < tmpListeners.size(); i++) {
-      if (tmpListeners.get(i).listener instanceof ParentListener) {
-        tmpListeners.get(i).deliver();
-      }
-    }
-    if (cancellableAncestor != null) {
-      cancellableAncestor.removeListener(parentListener);
-    }
+    cancellableAncestor.removeListener(cancellationListener);
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply context/src/main/java/io/grpc/Context.java]: Good catch, added test to confirm it works.

**Comment by bogdandrutu** [comment]
>  For example, "Reduce the size of the Context object and improve performance for operations like withValue" could be solved with lazy allocation of the parentListener. Yes, the two pointers will still be there, but a trivial change could yield large benefits. 

See https://github.com/grpc/grpc-java/pull/6951

**Comment by ejona86** [comment]
@bogdandrutu, can you please express your objective? It seems like you are hoping that I infer that I was correct that it was for memory usage.

#6951 conflicts with this one, so we will need to choose just one.

**Comment by bogdandrutu** [comment]
@ejona86 sorry for the "race condition", I was updating the motivation of this PR when you wrote your comment. I really liked your suggestions to do more trivial changes, so I started to split other PRs from this one like #6951 + (I will send another small one related to removing the ParentListener class). I would still like to have this PR merged after if possible because I think having the separation of concerns between Context and CancellableContext benefits us (we can even make Context an interface / abstract class and provide alternative implementations, etc.).

**Comment by ejona86** [comment]
I see that you've now incorporated the changes from #6951 into this, so I follow along now. I merged #6951 but now this one will need to have the conflict resolved.

**Comment by ejona86** [comment]
Thank you!

**Pull request merged**